File: src/main/js/formulas/math/ISEVEN.ts
import { SheetColumnRowKey } from "../../models/common/SheetColumnRowKey";
import { Complex } from "../../models/common/Types";
import { Converters } from "../../utils/Converters";
import { Numbers } from "../../utils/Numbers";
import { AbstractFormula } from "../AbstractFormula";
import { FormulaName } from "../FormulaName";
import { ROUNDDOWN } from "./ROUNDDOWN";

export class ISEVEN extends AbstractFormula {
  static SELF: ISEVEN = new ISEVEN();
  NAME = FormulaName.ISEVEN;

  internal(origin: SheetColumnRowKey, ...values: Array<Complex>) {
    AbstractFormula.checkLength(values.length, 1, this.NAME);
    const value = Converters.toNumber(Converters.first(this.collateralLookup(origin, values[0])));
    const rounded = Converters.toNumber(ROUNDDOWN.SELF.run(null, value));
    return Numbers.isZero(rounded % 2);
  }
}

File: src/main/java/io/protobase/f7/formulas/math/ISEVEN.java
package io.protobase.f7.formulas.math;

import io.protobase.f7.formulas.AbstractFormula;
import io.protobase.f7.formulas.FormulaName;
import io.protobase.f7.models.GridColumnRowKey;
import io.protobase.f7.utils.Converters;

import java.util.function.BiFunction;
import java.util.function.Function;

public class ISEVEN extends AbstractFormula {
  public static FormulaName NAME = FormulaName.ISEVEN;
  public static ISEVEN SELF = new ISEVEN();

  public ISEVEN() {
    super();
  }

  public ISEVEN(Function<Object, Object> lookup, BiFunction<GridColumnRowKey, Object, Object> collateralLookup) {
    super(lookup, collateralLookup);
  }

  @Override
  public Object internal(GridColumnRowKey origin, Object... values) {
    checkLength(values.length, 1, NAME);
    double number = Converters.toDouble(Converters.first(collateralLookup.apply(origin, values[0])));
    double rounded = Converters.toDouble(ROUNDDOWN.SELF.apply(null, number));
    double remainder = rounded % 2;
    return remainder == 0 || remainder == -0;
  }
}
